src/network/networkd-conf.c

```diff
@@ -0,0 +1,155 @@
+        [DUID_TYPE_RAW]  = "raw",
+        [DUID_TYPE_LLT]  = "link-layer-time",
+        [DUID_TYPE_EN]   = "vendor",
+        [DUID_TYPE_LL]   = "link-layer",
+        [DUID_TYPE_UUID] = "uuid"
+};
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING(duid_type, DUIDType);
+DEFINE_CONFIG_PARSE_ENUM(config_parse_duid_type, duid_type, DUIDType, "Failed to parse DUID type");
+
+int config_parse_duid_rawdata(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        int r;
+        long byte;
+        char *cbyte, *pnext;
+        const char *pduid = rvalue;
+        size_t count = 0, duid_index = 0;
+        Manager *m;
+        Network *n;
+        DUIDType *duid_type;
+        uint16_t *dhcp_duid_type;
+        size_t *dhcp_duid_len;
+        uint8_t *dhcp_duid;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        if (ltype == DUID_CONFIG_SOURCE_GLOBAL) {
+                m = userdata;
+                duid_type = &m->duid_type;
+                dhcp_duid_type = &m->dhcp_duid_type;
+                dhcp_duid_len = &m->dhcp_duid_len;
+                dhcp_duid = m->dhcp_duid;
+        } else {
+                /* DUID_CONFIG_SOURCE_NETWORK */
+                n = userdata;
+                duid_type = &n->duid_type;
+                dhcp_duid_type = &n->dhcp_duid_type;
+                dhcp_duid_len = &n->dhcp_duid_len;
+                dhcp_duid = n->dhcp_duid;
+        }
+
+        if (*duid_type == _DUID_TYPE_INVALID)
+                *duid_type = DUID_TYPE_RAW;
+
+        switch (*duid_type) {
+        case DUID_TYPE_LLT:
+                /* RawData contains DUID-LLT link-layer address (offset 6) */
+                duid_index = 6;
+                break;
+        case DUID_TYPE_EN:
+                /* RawData contains DUID-EN identifier (offset 4) */
+                duid_index = 4;
+                break;
+        case DUID_TYPE_LL:
+                /* RawData contains DUID-LL link-layer address (offset 2) */
+                duid_index = 2;
+                break;
+        case DUID_TYPE_UUID:
+                /* RawData specifies UUID (offset 0) - fall thru */
+        case DUID_TYPE_RAW:
+                /* First two bytes of RawData is DUID Type - fall thru */
+        default:
+                break;
+        }
+
+        if (*duid_type != DUID_TYPE_RAW)
+                *dhcp_duid_type = (uint16_t)(*duid_type);
+
+        /* RawData contains DUID in format " NN:NN:NN... " */
+        while (true) {
+                r = extract_first_word(&pduid, &cbyte, ":", 0);
+                if (r < 0) {
+                        log_error("Failed to read DUID.");
```

Comment:
please use log_syntax() for logging configuration file parsing errors. See the other parser calls. Also, invalid input should generally not result in total failure, but just warnings, for compat reasons. Only things like OOM and so on should result in total failure.